2-completed-app/webapp/helper.py

```diff
@@ -0,0 +1,115 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import json
+import os
+
+from azure.data.tables import TableServiceClient
+from dotenv import load_dotenv
+from flask import request
+
+load_dotenv(verbose=True)
+
+
+class TableServiceHelper:
+
+    def __init__(self, table_name=None, conn_str=None):
+        self.table_name = table_name if table_name else os.getenv("table_name")
+        self.conn_str = conn_str if conn_str else os.getenv("conn_str")
+        self.table_service = TableServiceClient.from_connection_string(self.conn_str)
+        self.table_client = self.table_service.get_table_client(self.table_name)
+
+    def query_entity(self, params):
+        filters = []
+        if params.get("partitionKey"):
+            filters.append("PartitionKey eq '{}'".format(params.get("partitionKey")))
+        if params.get("rowKeyDateStart") and params.get("rowKeyTimeStart"):
+            filters.append("RowKey ge '{} {}'".format(params.get("rowKeyDateStart"), params.get("rowKeyTimeStart")))
+        if params.get("rowKeyDateEnd") and params.get("rowKeyTimeEnd"):
+            filters.append("RowKey le '{} {}'".format(params.get("rowKeyDateEnd"), params.get("rowKeyTimeEnd")))
+        if params.get("minTemperature"):
+            filters.append("Temperature ge {}".format(params.get("minTemperature")))
+        if params.get("maxTemperature"):
+            filters.append("Temperature le {}".format(params.get("maxTemperature")))
+        if params.get("minPrecipitation"):
+            filters.append("Precipitation ge {}".format(params.get("minPrecipitation")))
+        if params.get("maxPrecipitation"):
+            filters.append("Precipitation le {}".format(params.get("maxPrecipitation")))
+        return list(self.table_client.query_entities(" and ".join(filters)))
+
+    def delete_entity(self):
+        partition_key = request.form.get("StationName")
+        row_key = request.form.get("ObservationDate")
+        return self.table_client.delete_entity(partition_key, row_key)
+
+    def insert_entity(self):
+        entity = self.deserialize()
+        return self.table_client.create_entity(entity)
+
+    def upsert_entity(self):
+        entity = self.deserialize()
```

Comment:
We shouldn't call a static method using `self`. It should be `TableServiceHelper.deserialize()`
Though it doesn't look like there's any reason that `deserialize`/`serialize` should be static?
And reason they couldn't just be standalone functions?

Reply:
Actually nevermind - looks like this should be fine :)